core/src/main/scala/com/salesforce/op/filters/RawFeatureFilter.scala

```diff
@@ -182,95 +191,68 @@ class RawFeatureFilter[T]
-
-    def logExcluded(excluded: Seq[Boolean], message: String): Unit = {
-      val featuresDropped = trainingDistribs.zip(excluded)
-        .collect{ case (f, d) if d => s"${f.name} ${f.key.getOrElse("")}" }
-      log.info(s"$message: ${featuresDropped.mkString(", ")}")
+    val allNumericReasons = getAllReasons(
+      trainingDistribs = trainingDistribs.numericDistributions,
+      scoringDistribs = scoringDistribs.map(_.numericDistributions).getOrElse(Map()),
+      correlationInfo = correlationInfo,
+      minFill = minFill,
+      maxCorrelation = maxCorrelation,
+      maxFillDifference = maxFillDifference,
+      maxFillRatioDiff = maxFillRatioDiff,
+      maxJSDivergence = maxJSDivergence,
+      jsDivergenceProtectedFeatures = jsDivergenceProtectedFeatures,
+      jsDivergenceF = FeatureDistribution.densityJSDivergence)
+    val allTextReasons = getAllReasons(
+      trainingDistribs = trainingDistribs.textDistributions,
+      scoringDistribs = scoringDistribs.map(_.textDistributions).getOrElse(Map()),
+      correlationInfo = correlationInfo,
+      minFill = minFill,
+      maxCorrelation = maxCorrelation,
+      maxFillDifference = maxFillDifference,
+      maxFillRatioDiff = maxFillRatioDiff,
+      maxJSDivergence = maxJSDivergence,
+      jsDivergenceProtectedFeatures = jsDivergenceProtectedFeatures,
+      jsDivergenceF = FeatureDistribution.densityJSDivergence)
+
+    val toDrop: Map[FeatureKey, (String, List[String])] = (allNumericReasons ++ allTextReasons).filter {
+      case ((name, _), (featureDescription, reasons)) =>
+        log.info(featureDescription)
+        reasons.nonEmpty && !protectedFeatures(name)
     }
+    val toDropKeySet: Set[FeatureKey] = toDrop.keySet
 
-    val featureSize = trainingDistribs.length
-
-    val trainingUnfilled = trainingDistribs.map(_.fillRate() < minFill)
-    logExcluded(trainingUnfilled, s"Features excluded because training fill rate did not meet min required ($minFill)")
-
-    val trainingNullLabelLeakers = {
-      if (correlationInfo.isEmpty) Seq.fill(featureSize)(false)
-      else {
-        val absoluteCorrs = correlationInfo.map(_._2)
-        for {distrib <- trainingDistribs} yield {
-          // Only filter if feature absolute null-label leakage correlation is greater than allowed correlation
-          val nullLabelLeakerIndicators = absoluteCorrs.map(_.get(distrib.featureKey).exists(_ > maxCorrelation))
-          nullLabelLeakerIndicators.exists(identity)
-        }
-      }
+    toDrop.foreach { case (featureKey, (_, reasons)) =>
+      log.info(s"Dropping feature $featureKey because:\n\t${reasons.mkString("\n\t")}\n")
     }
-    logExcluded(
-      trainingNullLabelLeakers,
-      s"Features excluded because null indicator correlation (absolute) exceeded max allowed ($maxCorrelation)")
-
-    val scoringUnfilled =
-      if (scoringDistribs.nonEmpty) {
-        require(scoringDistribs.length == featureSize, "scoring and training features must match")
```

Comment:
what will happen if scoring Distribution is empty of feature size does not match now, is it checked somewhere?